### ISSUE TITLE
Compiler should catch the arguments type mismatch in generic functions

### ISSUE THREAD

**Issue by jozefchutka** [microsoft/TypeScript#44410]
# Bug Report

Compiler should be stricter for generic types.

### 🔎 Search Terms

- generics
- type checking

### 🕗 Version & Regression Information

All versions including v4.3.2

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/MYGwhgzhAECC0G8C+BYAUKSMBCjVoDMBXAO2ABcBLAexOnIFMJyAeAFQD4AKAfQDcAjAC42AGmj8ATCICUedI2ZcADOPIAnIgxkBuaAHp9cdQHMiAWwYly0agXoBPAA4NoAcmbrKJE2+iUYEmobLEoTEjAAIxBXcmpoJzB1MEtGdVt7cmdXNxILSIZ1NwA6BSZyLg0tcRIGAHc4LhldAyMIAAtqIhAAE3p29WoG4GpzJ0pwKlpoQsH1MqVahtgmtU1tPUNoDq7e-sHh0fHJmjpZ6nm0RQqlxplxWBatne6+8gGh6BGxibAps-UcwWN3qdxqoOwTSebU6r32n2+xz+pxmgIu6CAA)

### 💻 Code

```ts
class A {}
class B {}
function test<T>(_v1:T, _v2:T) {}
test(0, true); // OK: Argument of type 'string' is not assignable to parameter of type 'number'.
test(true, new A()); // should throw compilation error
test(new A(), true); // should throw compilation error
test(new A(), A); // should throw compilation error
test(new A(), new B()); // should throw compilation error
```

### 🙁 Actual behavior

Typescript happily compiles various combinations of arguments, although differs in type.

### 🙂 Expected behavior

Typescript should be more strict to check generic types of arguments and fail in compiling when differs.

**Comment by RyanCavanaugh** [comment]
https://github.com/Microsoft/TypeScript/wiki/FAQ#why-do-these-empty-classes-behave-strangely

**Comment by jozefchutka** [comment]
Thanks @RyanCavanaugh for pointing me into the FAQ section. I am curious if this a temporary issue or intended and will not be fixed? I am aware of other programming languages that do recognize empty classes properly. Even in javascript, there are differences:

```
class A {
    constructor() {
        console.log(this.constructor.name);
    }
}

class B extends A {}

new A(); // prints "A"
new B(); // prints "B"
A === B // false
```

so even if structurally identical, it does behave differently in "runtime" and so the compile time type checking (purpose of typescript) is no help here. Is this something Typescript team will look into in future?

**Comment by fatcerberus** [comment]
No, the entire type system in TS is structural so this is intended behavior.  Classes IMO should have been nominally typed but they aren’t, they are structurally typed like everything else.

However, if you really need nominal typing, you can just add a private property to your class as a brand. Private properties are only considered structurally compatible if they come from the same class definition.
